src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -592,49 +598,65 @@ fn env_args(
-    // instead to avoid applying the same RUSTFLAGS to multiple targets
-    // arches:
+    let target_applies_to_host_kind = config.cli_unstable().target_applies_to_host_kind;
+    // Include untargeted configuration sources (like `RUSTFLAGS`) if
     //
-    // 1) If --target is not specified we just apply RUSTFLAGS to
-    // all builds; they are all going to have the same target.
+    //  - we're compiling artifacts for the target platform; or
+    //  - we're not cross-compiling; or
+    //  - we're compiling host artifacts, the requested target matches the host, and the user has
+    //    requested that the host pick up target configurations.
     //
-    // 2) If --target *is* specified then we only apply RUSTFLAGS
-    // to compilation units with the Target kind, which indicates
-    // it was chosen by the --target flag.
+    // The rationale for the third condition is that `RUSTFLAGS` is intended to affect compilation
+    // for the target, and `target-applies-to-host-kind` makes it so that the host is affected by
+    // its target's config, so if `--target <host triple>` then `RUSTFLAGS` should also apply to
+    // the host.
+    let include_generic = !kind.is_host()
+        || requested_kinds == [CompileKind::Host]
+        || (target_applies_to_host_kind
+            && requested_kinds == [CompileKind::Target(CompileTarget::new(host_triple)?)]);
+    // Include targeted configuration sources (like `target.*.rustflags`) if
     //
-    // This means that, e.g., even if the specified --target is the
-    // same as the host, build scripts in plugins won't get
-    // RUSTFLAGS.
-    if requested_kinds != [CompileKind::Host] && kind.is_host() {
-        // This is probably a build script or plugin and we're
-        // compiling with --target. In this scenario there are
-        // no rustflags we can apply.
-        return Ok(Vec::new());
-    }
-
-    // First try CARGO_ENCODED_RUSTFLAGS from the environment.
-    // Prefer this over RUSTFLAGS since it's less prone to encoding errors.
-    if let Ok(a) = env::var(format!("CARGO_ENCODED_{}", name)) {
-        if a.is_empty() {
-            return Ok(Vec::new());
+    //  - we're compiling artifacts for the target platform; or
+    //  - we're not cross-compiling; or
+    //  - we're compiling host artifacts and the user has requested that the host pick up target
+    //    configurations.
+    //
+    // The middle condition here may seem counter-intuitive. If `target-applies-to-host-kind` is
+    // disabled, then `target.*.rustflags` should arguably not apply to host artifacts. However,
+    // this is likely surprising to users who set `target.<host triple>.rustflags` and run `cargo
+    // build` (with no `--target`). So, we respect `target.<host triple>`.rustflags` when
+    // `--target` _isn't_ supplied, which arguably matches the mental model established by
+    // respecting `RUSTFLAGS` when `--target` isn't supplied.
```

Comment:
Why are you special casing this behavior for only rustflags? If you're doing a build with no `--target` passed and want target rustflags to automatically apply you could just do a regular `target-applies-to-host = true` default in that case and make sure rustflags respects that, special casing applying target rustflags to the host build separately from stuff like target linkers doesn't seem to make sense.

Maybe also have it so that if there is a host table present with no `--target` passed just automatically switch to `target-applies-to-host = false` default and only apply values from the host table.